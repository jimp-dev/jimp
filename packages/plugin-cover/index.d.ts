import { IllformedPlugin, ImageCallback } from '@jimp/core';

interface Cover extends IllformedPlugin {
  cover(w: number, h: number, cb?: ImageCallback): this;
  cover(w: number, h: number, alignBits?: number, cb?: ImageCallback): this;
  cover(
    w: number,
    h: number,
    alignBits?: number,
    mode?: string,
    cb?: ImageCallback
  ): this;
}

export default function(): Cover;
