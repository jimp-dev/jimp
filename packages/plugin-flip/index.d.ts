import { IllformedPlugin, ImageCallback } from '@jimp/core';

interface Flip extends IllformedPlugin {
  flip(horizontal: boolean, vertical: boolean, cb?: ImageCallback): this;
  mirror(horizontal: boolean, vertical: boolean, cb?: ImageCallback): this;
}

export default function(): Flip;
