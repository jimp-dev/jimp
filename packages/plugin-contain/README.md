# @jimp/plugin-contain

contain an image.
