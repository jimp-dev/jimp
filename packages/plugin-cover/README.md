<div align="center">
  <a href="https://intuit.github.io/Ignite/">
    <img width="200" height="200"
      src="https://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-11/256/crayon.png">
  </a>
  <h1>@jimp/plugin-cover</h1>
</div>

Scale the image so the given width and height keeping the aspect ratio. Some parts of the image may be clipped.

## Usage

- @param {number} w the width to resize the image to
- @param {number} h the height to resize the image to
- @param {number} alignBits (optional) A bitmask for horizontal and vertical alignment
- @param {string} mode (optional) a scaling method (e.g. Jimp.RESIZE_BEZIER)
- @param {function(Error, Jimp)} cb (optional) a callback for when complete

```js
import jimp from 'jimp';

async function main() {
  const image = await jimp.read('test/image.png');

  image.cover(150, 100);
}

main();
```
