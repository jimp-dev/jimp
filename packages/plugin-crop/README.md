<div align="center">
  <a href="https://intuit.github.io/Ignite/">
    <img width="200" height="200"
      src="https://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-11/256/crayon.png">
  </a>
  <h1>@jimp/plugin-crop</h1>
  <p>Crop an image.</p>
</div>

## Crop

Crops the image at a given point to a give size

- @param {number} x the x coordinate to crop form
- @param {number} y the y coordinate to crop form
- @param w the width of the crop region
- @param h the height of the crop region
- @param {function(Error, Jimp)} cb (optional) a callback for when complete

```js
import jimp from 'jimp';

async function main() {
  const image = await jimp.read('test/image.png');

  image.crop(150, 150);
}

main();
```

## AutoCrop

AutoCrop same color borders from this image

_ @param {number} tolerance (optional): a percent value of tolerance for pixels color difference (default: 0.0002%)
_ @param {boolean} cropOnlyFrames (optional): flag to crop only real frames: all 4 sides of the image must have some border (default: true)

```js
import jimp from 'jimp';

async function main() {
  const image = await jimp.read('test/image.png');

  image.autocrop();
}

main();
```
