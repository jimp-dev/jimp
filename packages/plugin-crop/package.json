{
  "name": "@jimp/plugin-crop",
  "version": "0.22.11",
  "description": "crop an image.",
  "main": "dist/index.js",
  "module": "es/index.js",
  "repository": "jimp-dev/jimp",
  "types": "index.d.ts",
  "scripts": {
    "test": "cross-env BABEL_ENV=test mocha --require @babel/register --recursive test  --extension js",
    "test:watch": "npm run test -- --reporter min --watch",
    "test:coverage": "nyc npm run test",
    "build": "npm run build:node:production && npm run build:module",
    "build:watch": "npm run build:node:debug -- -- --watch --verbose",
    "build:debug": "npm run build:node:debug",
    "build:module": "cross-env BABEL_ENV=module babel src -d es --source-maps --config-file ../../babel.config.js",
    "build:node": "babel src -d dist --source-maps --config-file ../../babel.config.js",
    "build:node:debug": "cross-env BABEL_ENV=development npm run build:node",
    "build:node:production": "cross-env BABEL_ENV=production npm run build:node"
  },
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@jimp/utils": "link:../utils"
  },
  "devDependencies": {
    "@jimp/custom": "link:../custom",
    "@jimp/test-utils": "link:../test-utils"
  },
  "peerDependencies": {
    "@jimp/custom": ">=0.3.5"
  },
  "publishConfig": {
    "access": "public"
  }
}
