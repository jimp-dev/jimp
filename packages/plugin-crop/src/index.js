/* eslint-disable no-labels */

import { throwError, isNodePattern } from '@jimp/utils';

export default function pluginCrop(event) {
  /**
   * Crops the image at a given point to a give size
   * @param {number} x the x coordinate to crop form
   * @param {number} y the y coordinate to crop form
   * @param w the width of the crop region
   * @param h the height of the crop region
   * @param {function(Error, Jimp)} cb (optional) a callback for when complete
   * @returns {Jimp} this for chaining of methods
   */
  event('crop', function(x, y, w, h, cb) {
    if (typeof x !== 'number' || typeof y !== 'number')
      return throwError.call(this, 'x and y must be numbers', cb);
    if (typeof w !== 'number' || typeof h !== 'number')
      return throwError.call(this, 'w and h must be numbers', cb);

    // round input
    x = Math.round(x);
    y = Math.round(y);
    w = Math.round(w);
    h = Math.round(h);

    if (x === 0 && w === this.bitmap.width) {
      // shortcut
      const start = (w * y + x) << 2;
      const end = (start + h * w) << (2 + 1);

      this.bitmap.data = this.bitmap.data.slice(start, end);
    } else {
      const bitmap = Buffer.allocUnsafe(w * h * 4);
      let offset = 0;

      this.scanQuiet(x, y, w, h, function(x, y, idx) {
        const data = this.bitmap.data.readUInt32BE(idx, true);
        bitmap.writeUInt32BE(data, offset, true);
        offset += 4;
      });

      this.bitmap.data = bitmap;
    }

    this.bitmap.width = w;
    this.bitmap.height = h;

    if (isNodePattern(cb)) {
      cb.call(this, null, this);
    }

    return this;
  });

  return {
    class: {
      /**
       * Autocrop same color borders from this image
       * @param {number} tolerance (optional): a percent value of tolerance for pixels color difference (default: 0.0002%)
       * @param {boolean} cropOnlyFrames (optional): flag to crop only real frames: all 4 sides of the image must have some border (default: true)
       * @param {function(Error, Jimp)} cb (optional): a callback for when complete (default: no callback)
       * @returns {Jimp} this for chaining of methods
       */
      autocrop(...args) {
        const w = this.bitmap.width;
        const h = this.bitmap.height;
        const minPixelsPerSide = 1; // to avoid cropping completely the image, resulting in an invalid 0 sized image

        let cb; // callback
        let tolerance = 0.0002; // percent of color difference tolerance (default value)
        let cropOnlyFrames = true; // flag to force cropping only if the image has a real "frame"
        // i.e. all 4 sides have some border (default value)

        // parse arguments
        for (let a = 0, len = args.length; a < len; a++) {
          if (typeof args[a] === 'number') {
            // tolerance value passed
            tolerance = args[a];
          }

          if (typeof args[a] === 'boolean') {
            // cropOnlyFrames value passed
            cropOnlyFrames = args[a];
          }

          if (typeof args[a] === 'function') {
            // callback value passed
            cb = args[a];
          }
        }

        /**
         * All borders must be of the same color as the top left pixel, to be cropped.
         * It should be possible to crop borders each with a different color,
         * but since there are many ways for corners to intersect, it would
         * introduce unnecessary complexity to the algorithm.
         */

        // scan each side for same color borders
        const colorTarget = this.getPixelColor(0, 0); // top left pixel color is the target color
        const rgba1 = this.constructor.intToRGBA(colorTarget);

        // for north and east sides
        let northPixelsToCrop = 0;
        let eastPixelsToCrop = 0;
        let southPixelsToCrop = 0;
        let westPixelsToCrop = 0;

        // north side (scan rows from north to south)
        north: for (let y = 0; y < h - minPixelsPerSide; y++) {
          for (let x = 0; x < w; x++) {
            const colorXY = this.getPixelColor(x, y);
            const rgba2 = this.constructor.intToRGBA(colorXY);

            if (this.constructor.colorDiff(rgba1, rgba2) > tolerance) {
              // this pixel is too distant from the first one: abort this side scan
              break north;
            }
          }
          // this row contains all pixels with the same color: increment this side pixels to crop
          northPixelsToCrop++;
        }

        // east side (scan columns from east to west)
        east: for (let x = 0; x < w - minPixelsPerSide; x++) {
          for (let y = 0 + northPixelsToCrop; y < h; y++) {
            const colorXY = this.getPixelColor(x, y);
            const rgba2 = this.constructor.intToRGBA(colorXY);

            if (this.constructor.colorDiff(rgba1, rgba2) > tolerance) {
              // this pixel is too distant from the first one: abort this side scan
              break east;
            }
          }
          // this column contains all pixels with the same color: increment this side pixels to crop
          eastPixelsToCrop++;
        }

        // south side (scan rows from south to north)
        south: for (
          let y = h - 1;
          y >= northPixelsToCrop + minPixelsPerSide;
          y--
        ) {
          for (let x = w - eastPixelsToCrop - 1; x >= 0; x--) {
            const colorXY = this.getPixelColor(x, y);
            const rgba2 = this.constructor.intToRGBA(colorXY);

            if (this.constructor.colorDiff(rgba1, rgba2) > tolerance) {
              // this pixel is too distant from the first one: abort this side scan
              break south;
            }
          }
          // this row contains all pixels with the same color: increment this side pixels to crop
          southPixelsToCrop++;
        }

        // west side (scan columns from west to east)
        west: for (
          let x = w - 1;
          x >= 0 + eastPixelsToCrop + minPixelsPerSide;
          x--
        ) {
          for (let y = h - 1; y >= 0 + northPixelsToCrop; y--) {
            const colorXY = this.getPixelColor(x, y);
            const rgba2 = this.constructor.intToRGBA(colorXY);

            if (this.constructor.colorDiff(rgba1, rgba2) > tolerance) {
              // this pixel is too distant from the first one: abort this side scan
              break west;
            }
          }
          // this column contains all pixels with the same color: increment this side pixels to crop
          westPixelsToCrop++;
        }

        // safety checks
        const widthOfPixelsToCrop = w - (westPixelsToCrop + eastPixelsToCrop);
        // widthOfPixelsToCrop >= 0 ? widthOfPixelsToCrop : 0;
        const heightOfPixelsToCrop =
          h - (southPixelsToCrop + northPixelsToCrop);
        // heightOfPixelsToCrop >= 0 ? heightOfPixelsToCrop : 0;

        // decide if a crop is needed
        let doCrop = false;

        if (cropOnlyFrames) {
          // crop image if all sides should be cropped
          doCrop =
            eastPixelsToCrop !== 0 &&
            northPixelsToCrop !== 0 &&
            westPixelsToCrop !== 0 &&
            southPixelsToCrop !== 0;
        } else {
          // crop image if at least one side should be cropped
          doCrop =
            eastPixelsToCrop !== 0 ||
            northPixelsToCrop !== 0 ||
            westPixelsToCrop !== 0 ||
            southPixelsToCrop !== 0;
        }

        if (doCrop) {
          // do the real crop
          this.crop(
            eastPixelsToCrop,
            northPixelsToCrop,
            widthOfPixelsToCrop,
            heightOfPixelsToCrop
          );
        }

        if (isNodePattern(cb)) {
          cb.call(this, null, this);
        }

        return this;
      }
    }
  };
}
