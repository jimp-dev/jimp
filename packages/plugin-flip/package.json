{
  "name": "@jimp/plugin-flip",
  "version": "0.22.11",
  "description": "flip an image.",
  "main": "dist/index.js",
  "module": "es/index.js",
  "repository": "jimp-dev/jimp",
  "types": "index.d.ts",
  "scripts": {
    "build": "npm run build:node:production && npm run build:module --recursive test",
    "build:watch": "npm run build:node:debug -- -- --watch --verbose",
    "build:debug": "npm run build:node:debug",
    "build:module": "cross-env BABEL_ENV=module babel src -d es --source-maps --config-file ../../babel.config.js",
    "build:node": "babel src -d dist --source-maps --config-file ../../babel.config.js",
    "build:node:debug": "cross-env BABEL_ENV=development npm run build:node",
    "build:node:production": "cross-env BABEL_ENV=production npm run build:node"
  },
  "author": "",
  "license": "MIT",
  "dependencies": {
    "@jimp/utils": "link:../utils"
  },
  "peerDependencies": {
    "@jimp/custom": ">=0.3.5",
    "@jimp/plugin-rotate": ">=0.3.5"
  },
  "publishConfig": {
    "access": "public"
  }
}
