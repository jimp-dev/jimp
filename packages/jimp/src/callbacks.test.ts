import { expect, test, describe } from "vitest";
import { makeTestImage } from "@jimp/test-utils";

import { HorizontalAlign, Jimp, VerticalAlign } from "./index.js";
import { CropOptions } from "@jimp/plugin-crop";
import { FlipOptions } from "@jimp/plugin-flip";
import { ResizeOptions, ScaleToFitOptions } from "@jimp/plugin-resize";
import { CoverOptions } from "@jimp/plugin-cover";
import { ContainOptions } from "@jimp/plugin-contain";

describe("Callbacks", () => {
  const targetImg = Jimp.fromBitmap(makeTestImage("▴▸▾", "◆▪▰", "▵▹▿"));
  const miniImg = Jimp.fromBitmap(makeTestImage("□▥", "▥■"));

  const operations = {
    crop: {
      args: [{ x: 1, y: 1, w: 2, h: 1 } as CropOptions],
    },
    invert: { args: [] },
    flip: {
      args: [{ horizontal: true, vertical: false } as FlipOptions],
    },
    gaussian: { args: [1] },
    blur: { args: [1] },
    greyscale: { args: [] },
    sepia: { args: [] },
    opacity: { args: [0.5] },
    resize: { args: [{ w: 2, h: 2 } as ResizeOptions] },
    scale: { args: [0.5] },
    brightness: { args: [1.5] },
    contrast: { args: [0.75] },
    posterize: { args: [5] },
    dither: { args: [] },
    // background: { args: [0xffffffff] },
    cover: {
      args: [
        {
          w: 3,
          h: 2,
          align: HorizontalAlign.LEFT | VerticalAlign.TOP,
        } as CoverOptions,
      ],
    },
    contain: {
      args: [
        {
          w: 3,
          h: 2,
          align: HorizontalAlign.LEFT | VerticalAlign.TOP,
        } as ContainOptions,
      ],
    },
    opaque: { args: [] },
    fade: { args: [0.5] },
    scaleToFit: { args: [{ w: 3, h: 2 } as ScaleToFitOptions] },
    blit: { args: [{ src: miniImg }] },
    composite: { args: [miniImg, 0, 0] },
  };

  Object.entries(operations).forEach(([op, { args }]) => {
    test("with " + op, () => {
      const clone = targetImg.clone();

      expect(
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        (clone[op as keyof typeof clone] as any)(...args)
      ).toMatchSnapshot();
    });
  });
});
