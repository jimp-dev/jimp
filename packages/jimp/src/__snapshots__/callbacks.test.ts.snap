// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Callbacks > with blit 1`] = `
Visualization:

□▥▾
▥■▰
▵▹▿

Data:

FF-FF-FFᶠᶠ BF-BF-BFᶠᶠ 00-00-FFᶠᶠ
BF-BF-BFᶠᶠ 00-00-00ᶠᶠ FF-00-FFᶠᶠ
FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
`;

exports[`Callbacks > with blur 1`] = `
Visualization:

7▩▵
F0F
2B4

Data:

AA-94-3Fᶠ¹ 7E-7E-7Eᶠ¹ 54-69-BFᶠ¹
9B-C2-52ᵈ⁴ 85-A7-85ᵈ⁴ 6E-8B-B6ᵈ⁴
88-00-6Cᵇ⁷ 8C-DC-8Cᵇ⁷ 92-B7-ACᵇ⁷
`;

exports[`Callbacks > with brightness 1`] = `
Visualization:

▴▸▾
◆▪▰
▵▹▿

Data:

FF-00-00ᶠᶠ 00-FF-00ᶠᶠ 00-00-FFᶠᶠ
FF-FF-00ᶠᶠ 00-FF-FFᶠᶠ FF-00-FFᶠᶠ
FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
`;

exports[`Callbacks > with composite 1`] = `
Visualization:

□▥▾
▥■▰
▵▹▿

Data:

FF-FF-FFᶠᶠ BF-BF-BFᶠᶠ 00-00-FFᶠᶠ
BF-BF-BFᶠᶠ 00-00-00ᶠᶠ FF-00-FFᶠᶠ
FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
`;

exports[`Callbacks > with contain 1`] = `
Visualization:

C0 
◇◇ 

Data:

AA-8E-1Cᶠᶠ 39-55-C6ᶠᶠ 00-00-00⁰⁰
71-5F-13ᵃᵃ 26-39-84ᵃᵃ 00-00-00⁰⁰
`;

exports[`Callbacks > with contrast 1`] = `
Visualization:

▴▸▾
◆▪▰
▵▹▿

Data:

FF-00-00ᶠᶠ 00-FF-00ᶠᶠ 00-00-FFᶠᶠ
FF-FF-00ᶠᶠ 00-FF-FFᶠᶠ FF-00-FFᶠᶠ
FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
`;

exports[`Callbacks > with cover 1`] = `
Visualization:

▴▸▾
◆▪▰

Data:

FF-00-00ᶠᶠ 00-FF-00ᶠᶠ 00-00-FFᶠᶠ
FF-FF-00ᶠᶠ 00-FF-FFᶠᶠ FF-00-FFᶠᶠ
`;

exports[`Callbacks > with crop 1`] = `
Visualization:

▪▰

Data:

00-FF-FFᶠᶠ FF-00-FFᶠᶠ
`;

exports[`Callbacks > with dither 1`] = `
Visualization:

▸FD
85▰
0▵▦

Data:

FF-01-01ᶠᶠ 09-FF-09ᶠᶠ 03-03-FFᶠᶠ
FF-FF-0Dᶠᶠ 05-FF-FFᶠᶠ FF-0F-FFᶠᶠ
FF-04-04⁷ᶠ 0C-FF-0C⁷ᶠ 02-02-FF⁷ᶠ
`;

exports[`Callbacks > with fade 1`] = `
Visualization:

▵▹▿
◇▫▱
3▹F

Data:

FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
FF-FF-00⁷ᶠ 00-FF-FF⁷ᶠ FF-00-FF⁷ᶠ
FF-00-00³ᶠ 00-FF-00³ᶠ 00-00-FF³ᶠ
`;

exports[`Callbacks > with flip 1`] = `
Visualization:

▾▸▴
▰▪◆
▿▹▵

Data:

00-00-FFᶠᶠ 00-FF-00ᶠᶠ FF-00-00ᶠᶠ
FF-00-FFᶠᶠ 00-FF-FFᶠᶠ FF-FF-00ᶠᶠ
00-00-FF⁷ᶠ 00-FF-00⁷ᶠ FF-00-00⁷ᶠ
`;

exports[`Callbacks > with gaussian 1`] = `
Visualization:

0▸7
▹▥7
▥▱F

Data:

72-A0-2Bᶠ⁸ 4F-70-92ᶠ⁵ 59-50-B9ᶠ³
78-7F-4Fᵈ⁴ 6F-64-79ᶜᶠ 45-55-A4ᶜᶜ
55-8F-51ᶜ³ 49-5A-8Fᵇᵉ 51-5D-93ᶜᶠ
`;

exports[`Callbacks > with greyscale 1`] = `
Visualization:

◇D▩
3▴D
▥81

Data:

36-36-36ᶠᶠ B6-B6-B6ᶠᶠ 12-12-12ᶠᶠ
EC-EC-ECᶠᶠ C8-C8-C8ᶠᶠ 48-48-48ᶠᶠ
36-36-36⁷ᶠ B6-B6-B6⁷ᶠ 12-12-12⁷ᶠ
`;

exports[`Callbacks > with invert 1`] = `
Visualization:

▪▰◆
▾▴▸
▫▱◇

Data:

00-FF-FFᶠᶠ FF-00-FFᶠᶠ FF-FF-00ᶠᶠ
00-00-FFᶠᶠ FF-00-00ᶠᶠ 00-FF-00ᶠᶠ
00-FF-FF⁷ᶠ FF-00-FF⁷ᶠ FF-FF-00⁷ᶠ
`;

exports[`Callbacks > with opacity 1`] = `
Visualization:

▵▹▿
◇▫▱
3▹F

Data:

FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
FF-FF-00⁷ᶠ 00-FF-FF⁷ᶠ FF-00-FF⁷ᶠ
FF-00-00³ᶠ 00-FF-00³ᶠ 00-00-FF³ᶠ
`;

exports[`Callbacks > with opaque 1`] = `
Visualization:

▴▸▾
◆▪▰
▴▸▾

Data:

FF-00-00ᶠᶠ 00-FF-00ᶠᶠ 00-00-FFᶠᶠ
FF-FF-00ᶠᶠ 00-FF-FFᶠᶠ FF-00-FFᶠᶠ
FF-00-00ᶠᶠ 00-FF-00ᶠᶠ 00-00-FFᶠᶠ
`;

exports[`Callbacks > with posterize 1`] = `
Visualization:

▴▸▾
◆▪▰
▵▹▿

Data:

FF-00-00ᶠᶠ 00-FF-00ᶠᶠ 00-00-FFᶠᶠ
FF-FF-00ᶠᶠ 00-FF-FFᶠᶠ FF-00-FFᶠᶠ
FF-00-00⁷ᶠ 00-FF-00⁷ᶠ 00-00-FF⁷ᶠ
`;

exports[`Callbacks > with resize 1`] = `
Visualization:

C0
▿▰

Data:

AA-8E-1Cᶠᶠ 39-55-C6ᶠᶠ
AA-8E-1Cᵃᵃ 39-55-C6ᵃᵃ
`;

exports[`Callbacks > with scale 1`] = `
Visualization:

C0
▿▰

Data:

AA-8E-1Cᶠᶠ 39-55-C6ᶠᶠ
AA-8E-1Cᵃᵃ 39-55-C6ᵃᵃ
`;

exports[`Callbacks > with scaleToFit 1`] = `
Visualization:

C0
▿▰

Data:

AA-8E-1Cᶠᶠ 39-55-C6ᶠᶠ
AA-8E-1Cᵃᵃ 39-55-C6ᵃᵃ
`;

exports[`Callbacks > with sepia 1`] = `
Visualization:

▥◇5
4□◆
6▥◇

Data:

64-22-2Dᶠᶠ C4-F3-B7ᶠᶠ 30-3B-4Eᶠᶠ
FF-FF-E5ᶠᶠ F4-FF-FFᶠᶠ 94-5E-7Cᶠᶠ
64-22-2D⁷ᶠ C4-F3-B7⁷ᶠ 30-3B-4E⁷ᶠ
`;
